### ISSUE TITLE
feat: RandomResizedCrop augmentation for ImageNet training

### ISSUE THREAD

**Issue by Samyssmile** [Samyssmile/edux#196]
## Summary

`RandomResizedCrop` is **the single most important augmentation** for ImageNet training. Every SOTA approach (ResNet, EfficientNet, ViT, ConvNeXt) relies on it. Without it, models see each image at the same scale and position every epoch, leading to massive overfitting.

**Estimated impact**: ~5–10% Top-1 accuracy improvement on ImageNet.

## Current State

The augmentation system only provides:
- `RandomHorizontalFlip` — horizontal mirroring
- `RandomBrightness` — uniform brightness offset
- `GaussianNoise` — per-pixel Gaussian noise

No spatial/geometric augmentations exist at all.

## What RandomResizedCrop Does

Given an input image of arbitrary size, it:
1. **Randomly samples a crop area** with configurable scale (e.g., 8%–100% of the image)
2. **Randomly samples an aspect ratio** (e.g., 3/4 to 4/3)
3. **Crops** the selected region from the source image
4. **Resizes** the crop to the target size (e.g., 224×224) using bilinear interpolation

```
Input Image (480×640)
┌──────────────────────────┐
│                          │
│    ┌──────────┐          │
│    │ Random   │          │    Resize
│    │ Crop     │  ──────────────────▶  [224 × 224]
│    │ (scale,  │          │            (bilinear)
│    │  ratio)  │          │
│    └──────────┘          │
│                          │
└──────────────────────────┘
```

### PyTorch Reference
```python
transforms.RandomResizedCrop(
    size=224,
    scale=(0.08, 1.0),       # crop area = 8%-100% of image
    ratio=(0.75, 1.3333),    # aspect ratio range
    interpolation=BILINEAR
)
```

## Proposed API

```java
public final class RandomResizedCrop implements ImageAugmentation {

    /**
     * @param targetHeight  output height (e.g., 224)
     * @param targetWidth   output width (e.g., 224)
     * @param minScale      minimum crop area ratio (default 0.08)
     * @param maxScale      maximum crop area ratio (default 1.0)
     * @param minRatio      minimum aspect ratio (default 0.75)
     * @param maxRatio      maximum aspect ratio (default 1.3333)
     */
    public RandomResizedCrop(int targetHeight, int targetWidth,
                              float minScale, float maxScale,
                              float minRatio, float maxRatio) { ... }

    // Convenience: common ImageNet defaults
    public RandomResizedCrop(int targetSize) {
        this(targetSize, targetSize, 0.08f, 1.0f, 0.75f, 1.3333f);
    }

    @Override
    public Tensor apply(Tensor image) { ... }  // [C, H, W] → [C, targetH, targetW]
}
```

### Usage
```java
var augmentation = AugmentationPipeline.of(
    new RandomResizedCrop(224),           // ← NEW
    new RandomHorizontalFlip(),
    new ColorJitter(0.4f, 0.4f, 0.4f, 0.1f)
);
var augmentedTrain = new AugmentedDataset(trainDataset, augmentation);
```

### Algorithm (matches PyTorch)

```java
@Override
public Tensor apply(Tensor image) {
    float[] src = image.data();
    int c = image.shape().get(0);
    int h = image.shape().get(1);
    int w = image.shape().get(2);
    float area = h * w;

    // 1. Try up to 10 random crops
    for (int attempt = 0; attempt < 10; attempt++) {
        float targetArea = area * uniform(minScale, maxScale);
        float aspectRatio = exp(uniform(log(minRatio), log(maxRatio)));

        int cropW = round(sqrt(targetArea * aspectRatio));
        int cropH = round(sqrt(targetArea / aspectRatio));

        if (cropW <= w && cropH <= h) {
            int x0 = randomInt(0, w - cropW);
            int y0 = randomInt(0, h - cropH);
            return cropAndResize(src, c, h, w, x0, y0, cropW, cropH,
                                 targetWidth, targetHeight);
        }
    }

    // 2. Fallback: center crop at min(h,w), then resize
    int size = Math.min(h, w);
    int x0 = (w - size) / 2;
    int y0 = (h - size) / 2;
    return cropAndResize(src, c, h, w, x0, y0, size, size,
                         targetWidth, targetHeight);
}
```

### Bilinear Interpolation for Resize

```java
private Tensor cropAndResize(float[] src, int c, int srcH, int srcW,
                              int cropX, int cropY, int cropW, int cropH,
                              int outW, int outH) {
    float[] result = new float[c * outH * outW];
    float scaleY = (float) cropH / outH;
    float scaleX = (float) cropW / outW;

    for (int ch = 0; ch < c; ch++) {
        for (int oy = 0; oy < outH; oy++) {
            float srcY = cropY + (oy + 0.5f) * scaleY - 0.5f;
            int y0 = clamp((int) floor(srcY), 0, srcH - 1);
            int y1 = clamp(y0 + 1, 0, srcH - 1);
            float wy = srcY - floor(srcY);

            for (int ox = 0; ox < outW; ox++) {
                float srcX = cropX + (ox + 0.5f) * scaleX - 0.5f;
                int x0 = clamp((int) floor(srcX), 0, srcW - 1);
                int x1 = clamp(x0 + 1, 0, srcW - 1);
                float wx = srcX - floor(srcX);

                float val = (1 - wy) * ((1 - wx) * pixel(src, ch, y0, x0, srcW)
                                       + wx * pixel(src, ch, y0, x1, srcW))
                          + wy * ((1 - wx) * pixel(src, ch, y1, x0, srcW)
                                 + wx * pixel(src, ch, y1, x1, srcW));

                result[ch * outH * outW + oy * outW + ox] = val;
            }
        }
    }
    return FloatTensor.of(Shape.of(c, outH, outW), result);
}
```

## Interaction with ImageFolderDataset

Currently `ImageFolderDataset` resizes images at load time to fixed dimensions. With `RandomResizedCrop`, the dataset should ideally store images at their **original resolution** (or a larger size like 256×256) and let the augmentation handle the final crop+resize to 224×224.

This may require:
- An option in `ImageFolderDataset` to load images at a "pre-crop" resolution (e.g., 256×256)
- Or storing raw image bytes and decoding on-the-fly

## Acceptance Criteria

- [ ] `RandomResizedCrop` implements `ImageAugmentation` interface
- [ ] Produces tensors of exact `[C, targetH, targetW]` shape
- [ ] Bilinear interpolation for quality resizing
- [ ] Fallback to center crop when random sampling fails after 10 attempts
- [ ] Aspect ratio sampled in log-space (uniform in log domain)
- [ ] Default parameters match PyTorch: scale=(0.08, 1.0), ratio=(3/4, 4/3)
- [ ] Thread-safe (uses `ThreadLocalRandom`)
- [ ] Unit test: output shape correctness
- [ ] Unit test: verify crop stays within image bounds
- [ ] Unit test: fallback path triggers for extreme aspect ratios
- [ ] Integration test: train small CNN with RRC and verify convergence improvement

## References

- [ResNet Paper (He et al., 2015)](https://arxiv.org/abs/1512.03385) — Section 3.4 describes random crop training
- [PyTorch RandomResizedCrop](https://pytorch.org/vision/stable/generated/torchvision.transforms.RandomResizedCrop.html)
- [Inception Paper (Szegedy et al., 2015)](https://arxiv.org/abs/1512.00567) — Introduced scale/aspect sampling

**Comment by Samyssmile** [comment]
Fixed: added input shape rank validation to RandomResizedCrop.apply() — commit 910ab90
